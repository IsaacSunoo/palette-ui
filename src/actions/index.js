export const setError = error => ({
  type: "SET_ERROR",
  error
});

export const setLoading = loading => ({
  type: "SET_LOADING",
  loading
});

export const setPalette = palette => ({
  type: "SET_PALETTE",
  palette
});

export const setLock = colorID => ({
  type: "SET_LOCK",
  colorID
});

export const setModal = status => ({
  type: "SET_MODAL",
  status
});

export const setProjects = projects => ({
  type: "SET_PROJECTS",
  projects
});
