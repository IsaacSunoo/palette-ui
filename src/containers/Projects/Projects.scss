.Projects {
  display: flex;
  flex-direction: column;
  background: #0e2c3e;

  .Create-Project-Container {
    display: flex;
    justify-content: space-between;
    align-items: center;
    background: #21222D;
    padding: .5em 1em;
    box-shadow: 0px 2px 2px 0px #231f20;

    h2 {
      margin: 0;
      font-family: $fontRobotoCond;
      color: #ffffff12;
      font-weight: 100;
      font-size: 1.8em;
      text-transform: uppercase;
    }

    .Create-Info {
      display: flex;
    }

    form {
      display: flex;

      input {
        position: relative;
        z-index: 9;
        border: none;
        right: -.5em;
        outline: none;
        padding: .5em 1em;
        font-size: 1em;
        border-top-left-radius: 3px;
        border-bottom-left-radius: 3px;
        background: #ffffff12;
        margin: 0;
        color: #ffffff8f;
        font-family: $fontRobotoCond;
      }

      button {
        font-family: $fontRobotoCond;
        border: none;
        background: #F18805;
        outline: none;
        color: #fff;
        font-size: .9em;
        font-weight: bold;
        border-top-right-radius: 5px;
        border-bottom-right-radius: 5px;
        text-transform: uppercase;
        z-index: 10;
        cursor: pointer;
        transition: all .3s ease;

        &:hover {
          background: #ff9d24;
        }
      }
    }

  }

  .Current-Projects {
    display: flex;
    justify-content: space-around;
    flex-wrap: wrap;
    padding: 1em 0em 2em 0em;
  }
}
