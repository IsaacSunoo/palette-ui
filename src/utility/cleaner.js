export const cleanProjectsPalettes = (projects, palettes) => {
  const cleanedData = projects.map(project => {
    const { id, name } = project;
    let projectData =  { id, name, palettes: [] }
    palettes.forEach(palette => {
      if(palette.project_id === project.id){
        projectData.palettes.push(palette)
      }
    });
    return projectData;
  });
  return cleanedData;
}
