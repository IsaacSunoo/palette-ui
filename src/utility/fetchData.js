export const fetchData = async (url, body) => {
  try {
    const response = await fetch(url, body)
    const data = await response.json()
    if(!response.ok) { throw new Error(`'Fetch Call Cannot Be Made' ${data}`)}
    return data;
  } catch (error) {
    return error;
  }
}
