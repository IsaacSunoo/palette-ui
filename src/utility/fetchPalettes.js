export const fetchPalettes = async (url) => {
  try {
    const response = await fetch(url)
    if(!response.ok) { throw new Error('Fetch Call Cannot Be Made')}
    const data = await response.json()
    return data;
  } catch (error) {
    console.log(error);
    // this.props.setError('Fetch Call Cannont')
  }
}
