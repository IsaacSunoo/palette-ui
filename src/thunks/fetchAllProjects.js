import * as actions from '../actions/index';
import { fetchData } from '../utility/fetchData';
import { cleanProjectsPalettes } from '../utility/cleaner';

export const fetchAllProjects = () => {
  return async (dispatch) => {
    try {
      dispatch(actions.setLoading(true));
      const allProjects = await fetchData(process.env.REACT_APP_BACKEND_URL + '/api/v1/projects');
      const allPalettes = await fetchData(process.env.REACT_APP_BACKEND_URL + '/api/v1/palettes');
      dispatch(actions.setLoading(false));
      if(!allProjects.ok){
        throw Error(allProjects.statusText)
      }
      const combinedData = cleanProjectsPalettes(allProjects, allPalettes)
      dispatch(actions.setProjects(combinedData));
    } catch (error) {
      dispatch(actions.setError(error.message));
    }
  }
}
