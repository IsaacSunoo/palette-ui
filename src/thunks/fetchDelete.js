import * as actions from '../actions/index';
import { fetchData } from '../utility/fetchData';
import { fetchOptions } from '../utility/fetchOptions';

export const fetchDelete = (url) => {
  return async (dispatch) => {
    try {
      dispatch(actions.setLoading(true));
      const options = await fetchOptions('DELETE', {});
      await fetchData(url, options);
      dispatch(actions.setLoading(false));
    } catch (error) {
      dispatch(actions.setError(error.message));
    }
  }
}
