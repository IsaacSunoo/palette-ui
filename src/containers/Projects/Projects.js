import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import ProjectInfo from '../ProjectInfo/ProjectInfo';

import * as actions from '../../actions';

import { fetchAllProjects } from '../../thunks/fetchAllProjects';
import { fetchOptions } from '../../utility/fetchOptions';
import { fetchData } from '../../utility/fetchData';

export class Projects extends Component {
  constructor() {
    super()
    this.state = {
      name: ''
    }
  }

  handleSendProject = async event => {
    event.preventDefault();
    const { fetchAllProjects } = this.props;
    const options = await fetchOptions('POST', this.state);
    const response = await fetchData('http://localhost:3001/api/v1/projects', options)
    fetchAllProjects();
  }

  handleChange = event => {
    const { value, name } = event.target;
    this.setState({
      [name]: value
    });
  }

  render() {
    const { allProjects } = this.props;

    return (
      <div className='Projects'>
        <div className='Create-Project-Container'>
          <h2>Project Information</h2>
          <section className='Create-Info'>
            <form onSubmit={this.handleSendProject}>
              <input type='text' onChange={this.handleChange}
                                 value={this.state.name}
                                 name='name'
                                 placeholder='Create A New Project'
                                 required
                     />
              <button>Save Project</button>
            </form>
          </section>
        </div>
        <div className='Current-Projects'>
          {
            allProjects.length &&
            allProjects.map((project, index) => {
              return <ProjectInfo key={project.name} {...project} palettes={project.palettes} id={project.id}/>
            })
          }
        </div>
      </div>
    )
  }
}

export const mapStateToProps = state => ({
  allProjects: state.allProjects,
});

export const mapDispatchToProps = dispatch => ({
  setPalette: data => dispatch(actions.setPalette(data)),
  fetchAllProjects: data => dispatch(fetchAllProjects(data)),
});

Projects.propTypes = {
  allProjects: PropTypes.array
}

export default connect(mapStateToProps, mapDispatchToProps)(Projects);
