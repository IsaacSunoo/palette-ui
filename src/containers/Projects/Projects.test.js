import React from 'react';
import { shallow } from 'enzyme';
import { Projects, mapStateToProps, mapDispatchToProps } from './Projects';

import { fetchOptions } from '../../utility/fetchOptions';
jest.mock('../../utility/fetchOptions');
import { fetchAllProjects } from '../../thunks/fetchAllProjects';
jest.mock('../../thunks/fetchAllProjects');
import { fetchData } from '../../utility/fetchData';
jest.mock('../../utility/fetchData');

import * as actions from '../../actions/index';

const mockAllProjects = [
  {
    id: 8,
    name: "Test Project",
    palettes: [
      {
        id: 15,
        name: "Mikes Pal",
        color_1: "#93125a",
        color_2: "#ffffff",
        color_3: "#f13470",
        color_4: "#80064e",
        color_5: "#423c52",
        project_id: 8,
      }
    ]
  }
]

describe('Projects', () => {

  describe('Projects Component', () => {
    let wrapper;
    let mockfetchAllProjects = jest.fn();
    let mockSetPalette = jest.fn();

    beforeEach(() => {
      wrapper = shallow(<Projects allProjects={mockAllProjects}
        fetchAllProjects={mockfetchAllProjects}
        setPalette={mockSetPalette}
      />)
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    });

    it('should have default state', () => {
      expect(wrapper.state()).toEqual({
        name: "",
      });
    });

    it('should call fetchOptions when handleSendProject is invoked with correct params', async () => {
      const instance = wrapper.instance();
      const event = { preventDefault: () => {}};
      await instance.handleSendProject(event);
      expect(fetchOptions).toHaveBeenCalledWith('POST', { name: '' });
    });

    it('should call fetchData when handleSendProject is invoked with the correct params', async () => {
      const mockUrl = 'http://localhost:3001/api/v1/projects';
      const instance = wrapper.instance();
      const event = { preventDefault: () => {}};
      await instance.handleSendProject(event);
      const mockOptions = await fetchOptions('POST', { name: '' });
      expect(fetchData).toHaveBeenCalledWith(mockUrl, mockOptions);
    });

    it('should call fetchAllProjects when handleSendProject is invoked with correct params', async () => {
      const instance = wrapper.instance();
      const event = { preventDefault: () => {}};
      await instance.handleSendProject(event);
      expect(mockfetchAllProjects).toHaveBeenCalled();
    });

    it.skip('should change state when handleChange is invoked', () => {
      const instance = wrapper.instance();
      const mockNameEvent = {
        target: { name: 'name', value: 'super palette project palette' }
      };
      const spy = jest.spyOn(instance, 'handleSendProject');
      const button = wrapper.find('.Form-Send');
      // wrapper.instance().handleChange(mockNameEvent);
      button.simulate('submit', mockNameEvent);
      expect(spy).toHaveBeenCalledWith(mockNameEvent);
    });

  });

  describe('mapStateToProps', () => {
    it('should return an object with a modalDisplay value', () => {
      const mockState = {
        allProjects: mockAllProjects,
        fakeState: "Fake String",
      }
      const expected = {
        allProjects: mockAllProjects,
      }

      const mappedProps = mapStateToProps(mockState)

      expect(mappedProps).toEqual(expected)
    })
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch for fetchAllProjects', () => {
      //Setup
      const mockData = [
        { color: '#fffff', locked: false },
        { color: '#00000', locked: false },
        { color: '#231f20', locked: false },
        { color: '#e6e6e6', locked: false },
        { color: '#cccccc', locked: false }
      ];
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchAllProjects(mockData);
      //Execution
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchAllProjects(mockData);
      //Expectation
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch for setPalette', () => {
      const mockData = []
      const mockDispatch = jest.fn();
      const actionToDispatch = actions.setPalette(mockData);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setPalette(mockData);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

  });

});
