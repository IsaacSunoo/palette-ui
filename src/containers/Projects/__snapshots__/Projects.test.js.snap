// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects Projects Component should match the snapshot 1`] = `
<div
  className="Projects"
>
  <div
    className="Create-Project-Container"
  >
    <h2>
      Project Information
    </h2>
    <section
      className="Create-Info"
    >
      <form
        className="Form-Send"
        onSubmit={[Function]}
      >
        <input
          name="name"
          onChange={[Function]}
          placeholder="Create A New Project"
          required={true}
          type="text"
          value=""
        />
        <button>
          Save Project
        </button>
      </form>
    </section>
  </div>
  <div
    className="Current-Projects"
  >
    <ConnectFunction
      id={8}
      key="Test Project"
      name="Test Project"
      palettes={
        Array [
          Object {
            "color_1": "#93125a",
            "color_2": "#ffffff",
            "color_3": "#f13470",
            "color_4": "#80064e",
            "color_5": "#423c52",
            "id": 15,
            "name": "Mikes Pal",
            "project_id": 8,
          },
        ]
      }
    />
  </div>
</div>
`;
